### PR TITLE
✨ Add support to set `allocation_pools` for subnet

### PR DESCRIPTION
**Pull request by dulek**
**What this PR does / why we need it**:
This commit adds API that allows users to set `allocations_pools` in the subnet created by CAPO. This allows the users to restrict the IP address ranges that will be allocated automatically by OpenStack when creating Machines.

Users can utilize this to reserve addresses for VIPs (virtual IPs) or special nodes that will have predefined addresses and will be created later.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [X] squashed commits
- if necessary:
  - [X] includes documentation
  - [X] adds unit tests

/hold

### CODE REVIEW

**Comment by EmilienM** [comment]
IMO life would be beautiful if we converted `NodeCIDR` into `NodeCIDRs` which is a list of subnet filters or something.

**Comment by mdbooth** [comment]
Another advantage of defining a SubnetSpec struct is that it would be extensible in v1beta1.

**Comment by jichenjc** [comment]
>>>This doesn't feel right. allocationPools feels like a property of a specific created subnet. This is going to break when we (hopefully) do IPI dual-stack.

this is useful in my openstack implementation especially 2 networks are in same CIDR with different allocation pool
I think it's not that helpful if we go IPI or everything under CAPO control, but it might be helpful for some customer
is using their existing env for CAPO (those resource under admin's input)

**Comment by MaysaMacedo** [comment]
> This doesn't feel right. `allocationPools` feels like a property of a specific created subnet. This is going to break when we (hopefully) do IPI dual-stack.
> 
> I was already wondering if `NodeCIDRs` should contain some kind of SubnetSpec rather than just a CIDR. It was already mentioned that this would allow specifying what kind of IPv6 address allocation to use. I would be much happier if `allocationPools` was a property of a SubnetSpec.
> 
> Thoughts? cc @MaysaMacedo

Just sharing what was discussed on the slack thread, having multiple SubnetSpec make sense to me, it would be good to name the field to something more representative of its content.

**Comment by EmilienM** [comment]
please rebase on main

**Comment by mdbooth** [comment]
@dulek I rebased this for you, btw. No conflicts.

**Comment by dulek** [comment]
Thanks!

File: pkg/cloud/services/networking/network_test.go [hunk 1]
```diff
@@ -423,6 +424,260 @@ func Test_ReconcileExternalNetwork(t *testing.T) {
...
+	fakeSubnetID := "d08803fc-2fa5-4179-b9d7-8c43d0af2fe6"
+	fakeCIDR := "10.0.0.0/24"
+	fakeNetworkID := "d08803fc-2fa5-4279-b9f7-8c45d0ff2fe6"
+	fakeDNS := "10.0.10.200"
+
+	tests := []struct {
+		name             string
+		openStackCluster *infrav1.OpenStackCluster
+		expect           func(m *mock.MockNetworkClientMockRecorder)
+		want             *infrav1.OpenStackCluster
```
Comment (mdbooth) [inline pkg/cloud/services/networking/network_test.go:443]: The test doesn't check `want`, and if you add a check it fails.

Comment (dulek) [reply pkg/cloud/services/networking/network_test.go]: Fixed!

**Comment by EmilienM** [comment]
@dulek is on vacation, I'll take care of rebase.

**Comment by EmilienM** [comment]
/cc MaysaMacedo
for double check

**Pull request merged**